vector/src/main/java/im/vector/app/features/userdirectory/UserListViewModel.kt

```diff
@@ -0,0 +1,262 @@
+import timber.log.Timber
+import java.util.concurrent.TimeUnit
+
+private typealias KnownUsersSearch = String
+private typealias DirectoryUsersSearch = String
+private typealias ContactUsersSearch = String
+
+class UserListViewModel @AssistedInject constructor(@Assisted
+                                                    initialState: UserListViewState,
+                                                    private val contactsDataSource: ContactsDataSource,
+                                                    private val session: Session)
+    : VectorViewModel<UserListViewState, UserListAction, UserListViewEvents>(initialState) {
+
+    private val knownUsersSearch = BehaviorRelay.create<KnownUsersSearch>()
+    private val directoryUsersSearch = BehaviorRelay.create<DirectoryUsersSearch>()
+    private val contactUsersSearch = BehaviorRelay.create<ContactUsersSearch>()
+
+    private var allContacts: List<MappedContact> = emptyList()
+    private var mappedContacts: List<MappedContact> = emptyList()
+
+    @AssistedInject.Factory
+    interface Factory {
+        fun create(initialState: UserListViewState): UserListViewModel
+    }
+
+    companion object : MvRxViewModelFactory<UserListViewModel, UserListViewState> {
+        override fun create(viewModelContext: ViewModelContext, state: UserListViewState): UserListViewModel? {
+            return when (viewModelContext) {
+                is ActivityViewModelContext -> {
+                    when (viewModelContext.activity<FragmentActivity>()) {
+                        is CreateDirectRoomActivity  -> (viewModelContext.activity<CreateDirectRoomActivity>()).userListViewModelFactory.create(state)
+                        is InviteUsersToRoomActivity -> (viewModelContext.activity<InviteUsersToRoomActivity>()).userListViewModelFactory.create(state)
+                        else                         -> error("Wrong activity or fragment")
+                    }
+                }
+                is FragmentViewModelContext -> (viewModelContext.fragment() as UserListFragment).userListViewModelFactory.create(state)
+                else                        -> error("Wrong activity or fragment")
+            }
+        }
+    }
+
+    init {
+        observeUsers()
+        loadContacts()
+    }
+
+    override fun handle(action: UserListAction) {
+        when (action) {
+            is UserListAction.SearchUsers          -> handleSearchUsers(action.value)
+            is UserListAction.ClearSearchUsers     -> handleClearSearchUsers()
+            is UserListAction.SelectPendingInvitee -> handleSelectUser(action)
+            is UserListAction.RemovePendingInvitee -> handleRemoveSelectedUser(action)
+            is UserListAction.OnlyBoundContacts    -> handleOnlyBoundContacts(action)
+        }.exhaustive
+    }
+
+    private fun handleSearchUsers(searchTerm: String) {
+        setState {
+            copy(searchTerm = searchTerm)
+        }
+        knownUsersSearch.accept(searchTerm)
```

Comment:
If i search for a query, then change during loading, will it cancel the current query to process the next one?

Reply:
Done.